src/vs/workbench/contrib/chat/browser/chatEditing/chatEditingTextModelChangeService.ts

```diff
@@ -0,0 +1,400 @@
+		},
+	});
+
+	private static readonly _pendingEditDecorationOptions = ModelDecorationOptions.register({
+		isWholeLine: true,
+		description: 'chat-pending-edit',
+		className: 'chat-editing-pending-edit',
+		minimap: {
+			position: MinimapPosition.Inline,
+			color: themeColorFromId(pendingRewriteMinimap)
+		}
+	});
+
+	private static readonly _atomicEditDecorationOptions = ModelDecorationOptions.register({
+		isWholeLine: true,
+		description: 'chat-atomic-edit',
+		className: 'chat-editing-atomic-edit',
+		minimap: {
+			position: MinimapPosition.Inline,
+			color: themeColorFromId(pendingRewriteMinimap)
+		}
+	});
+
+	private _isEditFromUs: boolean = false;
+	public get isEditFromUs() {
+		return this._isEditFromUs;
+	}
+	private _allEditsAreFromUs: boolean = true;
+	public get allEditsAreFromUs() {
+		return this._allEditsAreFromUs;
+	}
+	private _diffOperation: Promise<IDocumentDiff | undefined> | undefined;
+	private _diffOperationIds: number = 0;
+
+	private readonly _diffInfo = observableValue<IDocumentDiff>(this, nullDocumentDiff);
+	public get diffInfo() {
+		return this._diffInfo.map(value => {
+			return {
+				...value,
+				originalModel: this.originalModel,
+				modifiedModel: this.modifiedModel,
+				keep: changes => this.applyEditFromUs(() => this._acceptHunk(changes)),
+				undo: changes => this.applyEditFromUs(() => this._rejectHunk(changes))
+			} satisfies IDocumentDiff2;
+		});
+	}
+
+	private readonly _editDecorationClear = this._register(new RunOnceScheduler(() => { this._editDecorations = this.modifiedModel.deltaDecorations(this._editDecorations, []); }, 500));
+	private _editDecorations: string[] = [];
+
+	private readonly didAcceptOrRejectAllHunks = new Emitter<'acceptedAllChanges' | 'rejectedAllChanges'>();
+	public readonly onDidAcceptOrRejectAllHunks = this.didAcceptOrRejectAllHunks.event;
+
+	private readonly didUserEditModel = new Emitter<void>();
+	public readonly onDidUserEditModel = this.didUserEditModel.event;
+
+	private _edit: StringEdit = StringEdit.empty;
+	public get edit() {
+		return this._edit;
+	}
+
+	constructor(
+		private readonly originalModel: ITextModel,
+		private readonly modifiedModel: ITextModel,
+		private readonly state: IObservable<ModifiedFileEntryState>,
+		@IEditorWorkerService private readonly _editorWorkerService: IEditorWorkerService,
+		@IAccessibilitySignalService private readonly _accessibilitySignalService: IAccessibilitySignalService,
+	) {
+		super();
+		this._register(this.modifiedModel.onDidChangeContent(e => {
+			this._mirrorEdits(e);
+		}));
+
+		this._register(toDisposable(() => {
+			this.clearCurrentEditLineDecoration();
+		}));
+	}
+
+	public clearCurrentEditLineDecoration() {
+		this._editDecorations = this.modifiedModel.deltaDecorations(this._editDecorations, []);
+	}
+
+	public async areOriginalAndModifiedIdentical(): Promise<boolean> {
+		const diff = await this._diffOperation;
+		return diff ? diff.identical : false;
```

Comment:
Curious why not just compare `originalModel.getValue() === modifiedModel.getValue()` or the like.